Modules/posixmodule.c

```diff
@@ -10073,8 +10073,27 @@ os__getdiskusage_impl(PyObject *module, path_t *path)
     Py_BEGIN_ALLOW_THREADS
     retval = GetDiskFreeSpaceExW(path->wide, &_, &total, &free);
     Py_END_ALLOW_THREADS
-    if (retval == 0)
+    if (retval == 0) {
+        if (GetLastError() == ERROR_DIRECTORY) {
+            PyObject *result = NULL;
+            wchar_t *wpath = NULL;
+            path_t dir_path = *path;
+
+            wpath = PyMem_New(wchar_t, dir_path.length + 1);
+            if (wpath == NULL)
+                return PyErr_NoMemory();
+
+            wcscpy_s(wpath, dir_path.length + 1, path->wide);
+            _dirnameW(wpath);
```

Comment:
`_dirnameW` is limited to `MAX_PATH`. Longer paths fail with -1 as the return code.

Reply:
Cool - I'll add the check, or do you mean I should not use `_dirnameW`, since in this case the path length can potentially exceed `MAX_PATH` characters? (I guess I'm a bit confused about Steve's comment, since parts of `posixmodule.c` still assume an upper bound of `MAX_PATH`, while others don't.)